### PR TITLE
Fix plan invalidation when data connection is changed/dropped

### LINKED ISSUE #24526: com.hazelcast.mapstore.mysql.MySQLGenericMapStoreTest.givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema [HZ-2426]
_master_ (commit 816c4499514e0c87f67720599d00f5e797a175b7)

This seems to happen constantly with MySQL 5.7.42. 

Failed on oracle-11 with MySQL 5.7.42: https://jenkins.hazelcast.com/view/JDBC/job/Hazelcast-master-mysql-5.7/3/testReport/com.hazelcast.mapstore.mysql/MySQLGenericMapStoreTest/givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema/
<details><summary>Stacktrace:</summary>

```
com.hazelcast.sql.HazelcastSqlException: From line 1, column 81 to line 1, column 84: Column 'id' not found in any table
	at com.hazelcast.sql.impl.CoreQueryUtils.toPublicException(CoreQueryUtils.java:50)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:221)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:172)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:168)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:164)
	at com.hazelcast.sql.SqlService.execute(SqlService.java:89)
	at com.hazelcast.mapstore.GenericMapLoader.load(GenericMapLoader.java:310)
	at com.hazelcast.mapstore.GenericMapLoaderTest.givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema(GenericMapLoaderTest.java:519)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.hazelcast.sql.impl.QueryException: From line 1, column 81 to line 1, column 84: Column 'id' not found in any table
	at com.hazelcast.sql.impl.QueryException.error(QueryException.java:69)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse(QueryParser.java:67)
	at com.hazelcast.jet.sql.impl.OptimizerContext.parse(OptimizerContext.java:155)
	at com.hazelcast.jet.sql.impl.CalciteSqlOptimizer.prepare(CalciteSqlOptimizer.java:306)
	at com.hazelcast.sql.impl.SqlServiceImpl.prepare(SqlServiceImpl.java:275)
	at com.hazelcast.sql.impl.SqlServiceImpl.query0(SqlServiceImpl.java:259)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:204)
	... 18 more
Caused by: org.apache.calcite.runtime.CalciteContextException: From line 1, column 81 to line 1, column 84: Column 'id' not found in any table
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.apache.calcite.runtime.Resources$ExInstWithCause.ex(Resources.java:505)
	at org.apache.calcite.sql.SqlUtil.newContextException(SqlUtil.java:932)
	at org.apache.calcite.sql.SqlUtil.newContextException(SqlUtil.java:917)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.newValidationError(HazelcastSqlValidator.java:517)
	at org.apache.calcite.sql.validate.DelegatingScope.fullyQualify(DelegatingScope.java:273)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$Expander.visit(SqlValidatorImpl.java:6511)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$Expander.visit(SqlValidatorImpl.java:6491)
	at org.apache.calcite.sql.SqlIdentifier.accept(SqlIdentifier.java:324)
	at org.apache.calcite.sql.util.SqlShuttle$CallCopyingArgHandler.visitChild(SqlShuttle.java:134)
	at org.apache.calcite.sql.util.SqlShuttle$CallCopyingArgHandler.visitChild(SqlShuttle.java:101)
	at org.apache.calcite.sql.SqlOperator.acceptCall(SqlOperator.java:954)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$Expander.visitScoped(SqlValidatorImpl.java:6531)
	at org.apache.calcite.sql.validate.SqlScopedShuttle.visit(SqlScopedShuttle.java:54)
	at org.apache.calcite.sql.validate.SqlScopedShuttle.visit(SqlScopedShuttle.java:37)
	at org.apache.calcite.sql.SqlCall.accept(SqlCall.java:161)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$Expander.go(SqlValidatorImpl.java:6500)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.expand(SqlValidatorImpl.java:6098)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateWhereClause(SqlValidatorImpl.java:4412)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelect(SqlValidatorImpl.java:3700)
	at org.apache.calcite.sql.validate.SelectNamespace.validateImpl(SelectNamespace.java:64)
	at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:89)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateNamespace(SqlValidatorImpl.java:1107)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateQuery(SqlValidatorImpl.java:1078)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.validateQuery(HazelcastSqlValidator.java:173)
	at org.apache.calcite.sql.SqlSelect.validate(SqlSelect.java:248)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateScopedExpression(SqlValidatorImpl.java:1053)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validate(SqlValidatorImpl.java:759)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.validate(HazelcastSqlValidator.java:168)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse0(QueryParser.java:80)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse(QueryParser.java:56)
	... 23 more
Caused by: org.apache.calcite.sql.validate.SqlValidatorException: Column 'id' not found in any table
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.apache.calcite.runtime.Resources$ExInstWithCause.ex(Resources.java:505)
	at org.apache.calcite.runtime.Resources$ExInst.ex(Resources.java:599)
	... 52 more

```
</details>

<details><summary>Standard output:</summary>

```
Finished Running Test: givenRow_whenLoadAll_thenReturnMapWithGenericRecord in 0.148 seconds.
Started Running Test: givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema
11:55:38,804 DEBUG || - [GenericMapLoader] hz.MySQLGenericMapStoreTest_ecstatic_haibt.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Initializing for map people_o_8be553dd_beaf_4466_81af_a90ca9f655eb
11:55:39,012  INFO |givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema| - [SqlTestSupport] Time-limited test - Removing 0 cached plans in SqlTestSupport.@After
11:55:39,012  INFO |givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema| - [SqlTestSupport] Time-limited test - Removing 0 cached plans in SqlTestSupport.@After
11:55:39,013  INFO |givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema| - [SimpleTestInClusterSupport] Time-limited test - Ditching 0 jobs in SimpleTestInClusterSupport.@After: []
11:55:39,013  INFO |givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema| - [SimpleTestInClusterSupport] Time-limited test - Destroying 1 distributed objects in SimpleTestInClusterSupport.@After: [hz:impl:mapService/__sql.catalog]
BuildInfo right after givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema(com.hazelcast.mapstore.mysql.MySQLGenericMapStoreTest): BuildInfo{version='5.3.0-SNAPSHOT', build='20230511', buildNumber=20230511, revision=816c449, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'givenTableNameProperty_whenCreateMapLoader_thenUseTableNameWithCustomSchema(com.hazelcast.mapstore.mysql.MySQLGenericMapStoreTest):'
11:55:35, accumulated pauses: 155 ms, max pause: 133 ms, pauses over 1000 ms: 0


No metrics recorded during the test

```
</details>

Standard output can be found here - https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-master-mysql-5.7/3/

### PR DESCRIPTION
**Pull request by frant-hartm**
The change in data connection is propagated to
InternalDataConnectionService only after the listeners for plan invalidation run. This workaround executes the listeners again after the change is propagated.

Fixes #24526
HZ-2426

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

**Review by TomaszGaweda — APPROVED** [review]

**Comment by k-jamroz** [comment]
Do we need to check if the data connection exists in `createTable`? If we need, this could be done somewhere around  `TableResolverImpl.toTable` so it applies to all connectors.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/DataConnectionResolver.java [hunk 1]
```diff
@@ -113,6 +113,13 @@ public void removeDataConnection(String name, boolean ifExists) {
         }
     }
 
+    /**
+     * Invoke all registerer change listeners
+     */
+    public void invokeChangeListeners() {
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/DataConnectionResolver.java:119]: maybe add TODO here as well, this method should not be public

Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/DataConnectionResolver.java:120]: 334545ba42f

**Comment by frant-hartm** [comment]
`extractConnector` uses the data connection to figure out the type of the connector, so we already need the data connection to exist. Note that I didn't need to add the check to `JdbcSqlConnector#createTable`

https://github.com/hazelcast/hazelcast/blob/eefb14a2b4f5718f9b632fe8bbc52e6662913520/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java#L308-L310

**Review by k-jamroz — APPROVED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcTable.java [hunk 2]
```diff
@@ -117,10 +124,52 @@ public List<String> getPrimaryKeyList() {
...
+                    && Objects.equals(tableName, that.tableName)
+                    && Arrays.equals(externalName, that.externalName)
+                    && Objects.equals(dataConnectionName, that.dataConnectionName)
+                    && Objects.equals(fields, that.fields)
+                    && Objects.equals(options, that.options);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = Objects.hash(schemaName, tableName, fields, options);
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcTable.java:170]: `dataConnectionName` missing

**Comment by frant-hartm** [comment]
Known test failure reported here https://github.com/hazelcast/hazelcast/issues/23761

**Pull request merged**